File: api/job_application/job_application.php
<?php

class JobApplication
{
    private $conn;
    private $table_name = "job_application_tb";

    public $application_id;
    public $job;
    public $applicant_first_name;
    public $applicant_last_name;
    public $degree;
    public $applicant_location;
    public $applicant_contact_no;
    public $applicant_email;
    public $resume;
    public $cover_letter;
    public $skills;
    public $certifications;
    public $application_status;
    public $date_applied;

    // job references job_posting_tb fields where the job is the job_id
    public $job_id;
    public $cover_photo;
    public $job_title;
    public $status;
    public $field_industry;
    public $job_level;
    public $yrs_of_experience_needed;
    public $contractual_status;
    public $salary;
    public $job_location;
    public $job_description;
    public $requirements;
    public $job_responsibilities;
    public $industry_partner;

    // Constructor with $db as database connection
    public function __construct($db)
    {
        $this->conn = $db;
    }

    public function create()
    {
        $query = "INSERT INTO " . $this->table_name . " (job, applicant_first_name, applicant_last_name, degree, applicant_location, applicant_contact_no, applicant_email, resume, cover_letter, skills, certifications, application_status, date_applied) 
              VALUES (:job, :applicant_first_name, :applicant_last_name, :degree, :applicant_location, :applicant_contact_no, :applicant_email, :resume, :cover_letter, :skills, :certifications, :application_status, :date_applied)";

        // Prepare query
        $stmt = $this->conn->prepare($query);

        // Sanitize input
        $this->job = htmlspecialchars(strip_tags($this->job));
        $this->applicant_first_name = htmlspecialchars(strip_tags($this->applicant_first_name));
        $this->applicant_last_name = htmlspecialchars(strip_tags($this->applicant_last_name));
        $this->degree = htmlspecialchars(strip_tags($this->degree));
        $this->applicant_location = htmlspecialchars(strip_tags($this->applicant_location));
        $this->applicant_contact_no = htmlspecialchars(strip_tags($this->applicant_contact_no));
        $this->applicant_email = htmlspecialchars(strip_tags($this->applicant_email));
        $this->resume = htmlspecialchars(strip_tags($this->resume));
        $this->cover_letter = htmlspecialchars(strip_tags($this->cover_letter));
        $this->skills = htmlspecialchars(strip_tags($this->skills));
        $this->certifications = htmlspecialchars(strip_tags($this->certifications));
        $this->application_status = htmlspecialchars(strip_tags($this->application_status));
        $this->date_applied = htmlspecialchars(strip_tags($this->date_applied));

        // Bind parameters
        $stmt->bindParam(":job", $this->job);
        $stmt->bindParam(":applicant_first_name", $this->applicant_first_name);
        $stmt->bindParam(":applicant_last_name", $this->applicant_last_name);
        $stmt->bindParam(":degree", $this->degree);
        $stmt->bindParam(":applicant_location", $this->applicant_location);
        $stmt->bindParam(":applicant_contact_no", $this->applicant_contact_no);
        $stmt->bindParam(":applicant_email", $this->applicant_email);
        $stmt->bindParam(":resume", $this->resume);
        $stmt->bindParam(":cover_letter", $this->cover_letter);
        $stmt->bindParam(":skills", $this->skills);
        $stmt->bindParam(":certifications", $this->certifications);
        $stmt->bindParam(":application_status", $this->application_status);
        $stmt->bindParam(":date_applied", $this->date_applied);

        // Execute query    
        if ($stmt->execute()) {
            return true;
        }
        return false;
    }

    public function read()
    {
        $query = "SELECT
                      ja.*,
                      jp.*
                  FROM " . $this->table_name . " ja
                  JOIN job_posting_tb jp ON ja.job = jp.job_id;";
        $stmt = $this->conn->prepare($query);
        $stmt->execute();
        return $stmt;
    }

    public function update()
    {
        $query = "UPDATE " . $this->table_name . " 
        SET
            job = :job,
            applicant_first_name = :applicant_first_name,
            applicant_last_name = :applicant_last_name,
            degree = :degree,
            applicant_location = :applicant_location,
            applicant_contact_no = :applicant_contact_no,
            applicant_email = :applicant_email,
            resume = :resume,
            cover_letter = :cover_letter,
            skills = :skills,
            certifications = :certifications,
            application_status = :application_status,
            date_applied = :date_applied
        WHERE application_id = :application_id";

        // Prepare query
        $stmt = $this->conn->prepare($query);

        // Sanitize input
        $this->application_id = htmlspecialchars(strip_tags($this->application_id));
        $this->job = htmlspecialchars(strip_tags($this->job));
        $this->applicant_first_name = htmlspecialchars(strip_tags($this->applicant_first_name));
        $this->applicant_last_name = htmlspecialchars(strip_tags($this->applicant_last_name));
        $this->degree = htmlspecialchars(strip_tags($this->degree));
        $this->applicant_location = htmlspecialchars(strip_tags($this->applicant_location));
        $this->applicant_contact_no = htmlspecialchars(strip_tags($this->applicant_contact_no));
        $this->applicant_email = htmlspecialchars(strip_tags($this->applicant_email));
        $this->resume = htmlspecialchars(strip_tags($this->resume));
        $this->cover_letter = htmlspecialchars(strip_tags($this->cover_letter));
        $this->skills = htmlspecialchars(strip_tags($this->skills));
        $this->certifications = htmlspecialchars(strip_tags($this->certifications));
        $this->application_status = htmlspecialchars(strip_tags($this->application_status));
        $this->date_applied = htmlspecialchars(strip_tags($this->date_applied));

        // Bind parameters
        $stmt->bindParam(":application_id", $this->application_id);
        $stmt->bindParam(":job", $this->job);
        $stmt->bindParam(":applicant_first_name", $this->applicant_first_name);
        $stmt->bindParam(":applicant_last_name", $this->applicant_last_name);
        $stmt->bindParam(":degree", $this->degree);
        $stmt->bindParam(":applicant_location", $this->applicant_location);
        $stmt->bindParam(":applicant_contact_no", $this->applicant_contact_no);
        $stmt->bindParam(":applicant_email", $this->applicant_email);
        $stmt->bindParam(":resume", $this->resume);
        $stmt->bindParam(":cover_letter", $this->cover_letter);
        $stmt->bindParam(":skills", $this->skills);
        $stmt->bindParam(":certifications", $this->certifications);
        $stmt->bindParam(":application_status", $this->application_status);
        $stmt->bindParam(":date_applied", $this->date_applied);
        
        // Execute query
        return $stmt->execute();
    }

    public function delete()
    {
        $query = "DELETE FROM " . $this->table_name . " WHERE application_id = ?";
        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(1, $this->application_id);
        return $stmt->execute();
    }
}

File: lib/models/recruitment_and_placement/job_application.dart

// Define the JobApplication class with job posting properties only

class JobApplication {
  final int? applicationId;
  final int job;
  final String applicantFirstName;
  final String applicantLastName;
  final String applicantLocation;
  final String applicantContactNo;
  final String applicantEmail;
  final String resume;
  final String coverLetter;
  final String skills;
  final String certifications;
  final String applicationStatus;
  final String dateApplied;

  // Constructor for job posting properties only
  JobApplication({
    this.applicationId,
    required this.job,
    required this.applicantFirstName,
    required this.applicantLastName,
    required this.applicantLocation,
    required this.applicantContactNo,
    required this.applicantEmail,
    required this.resume,
    required this.coverLetter,
    required this.skills,
    required this.certifications,
    required this.applicationStatus,
    required this.dateApplied,
  });

  // Factory method for creating a Job Application from JSON
  factory JobApplication.fromJson(Map<String, dynamic> json) {
    return JobApplication(
      applicationId: json['application_id'] as int?,
      job: json['job'] as int? ?? 0,
      applicantFirstName: json['applicant_first_name'] ?? '',
      applicantLastName: json['applicant_last_name'] ?? '',
      applicantLocation: json['applicant_location'] ?? '',
      applicantContactNo: json['applicant_contact_no'] ?? '',
      applicantEmail: json['applicant_email'] ?? '',
      resume: json['resume'] ?? '',
      coverLetter: json['cover_letter'] ?? '',
      skills: json['skills'] ?? '',
      certifications: json['certifications'] ?? '',
      applicationStatus: json['application_status'] ?? '',
      dateApplied: json['date_applied'] ?? '',
    );
  }

  // Method to convert JobApplication to JSON
  Map<String, dynamic> toJson() {
    return {
      'application_id': applicationId,
      'job': job,
      'applicant_first_name': applicantFirstName,
      'applicant_last_name': applicantLastName,
      'applicant_location': applicantLocation,
      'applicant_contact_no': applicantContactNo,
      'applicant_email': applicantEmail,
      'resume': resume,
      'cover_letter': coverLetter,
      'skills': skills,
      'certifications': certifications,
      'application_status': applicationStatus,
      'date_applied': dateApplied,
    };
  }
}

// Define JobApplicationComplete class, extending JobApplication
class JobApplicationComplete extends JobApplication {

  // Properties for job_posting_tb
  final int? jobId;
  final String coverPhoto;
  final String jobTitle;
  final String status;
  final String fieldIndustry;
  final String jobLevel;
  final String yrsOfExperienceNeeded;
  final String contractualStatus;
  final String salary;
  final String jobLocation;
  final String jobDescription;
  late final String requirements;
  late final String jobResponsibilities;
  final int industryPartner;

  // Constructor for job posting with industry partner properties
  JobApplicationComplete({
    // JobApplication properties
    super.applicationId,
    required super.job,
    required super.applicantFirstName,
    required super.applicantLastName,
    required super.applicantLocation,
    required super.applicantContactNo,
    required super.applicantEmail,
    required super.resume,
    required super.coverLetter,
    required super.skills,
    required super.certifications,
    required super.applicationStatus,
    required super.dateApplied,

    // Job_posting_tb properties
    this.jobId,
    required this.coverPhoto,
    required this.jobTitle,
    required this.status,
    required this.fieldIndustry,
    required this.jobLevel,
    required this.yrsOfExperienceNeeded,
    required this.contractualStatus,
    required this.salary,
    required this.jobLocation,
    required this.jobDescription,
    required this.requirements,
    required this.jobResponsibilities,
    required this.industryPartner,
  });

  // Factory method for creating JobApplicationComplete from JSON
  factory JobApplicationComplete.fromJson(Map<String, dynamic> json) {
    return JobApplicationComplete(
      // JobApplication properties
      applicationId: json['application_id'] as int?,
      job: json['job'] ?? 0,
      applicantFirstName: json['applicant_first_name'] ?? '',
      applicantLastName: json['applicant_last_name'] ?? '',
      applicantLocation: json['applicant_location'] ?? '',
      applicantContactNo: json['applicant_contact_no'] ?? '',
      applicantEmail: json['applicant_email'] ?? '',
      resume: json['resume'] ?? '',
      coverLetter: json['cover_letter'] ?? '',
      skills: json['skills'] ?? '',
      certifications: json['certifications'] ?? '',
      applicationStatus: json['application_status'] ?? '',
      dateApplied: json['date_applied'] ?? '',

      // Job_posting_tb properties
      jobId: json['job_id'] as int? ?? 0,
      coverPhoto: json['cover_photo'] ?? '',
      jobTitle: json['job_title'] ?? '',
      status: json['status'] ?? '',
      fieldIndustry: json['field_industry'] ?? '',
      jobLevel: json['job_level'] ?? '',
      yrsOfExperienceNeeded: json['yrs_of_experience_needed'] ?? '',
      contractualStatus: json['contractual_status'],
      salary: json['salary'] ?? '',
      jobLocation: json['job_location'] ?? '',
      jobDescription: json['job_description'] ?? '',
      requirements: json['requirements'] ?? '',
      jobResponsibilities: json['job_responsibilities'] ?? '',
      industryPartner: json['industry_partner'] as int,
    );
  }

  // Method to convert JobApplicationComplete to JSON
  @override
  Map<String, dynamic> toJson() {
    final json = super.toJson();
    json.addAll({
      // Job_posting_tb properties
      'job_id': jobId,
      'cover_photo': coverPhoto,
      'job_title': jobTitle,
      'status': status,
      'field_industry': fieldIndustry,
      'job_level': jobLevel,
      'yrs_of_experience_needed': yrsOfExperienceNeeded,
      'contractual_status': contractualStatus,
      'salary': salary,
      'job_location': jobLocation,
      'job_description': jobDescription,
      'requirements': requirements,
      'job_responsibilities': jobResponsibilities,
      'industry_partner': industryPartner,
    });
    return json;
  }
}
